clang/test/SemaSYCL/max-concurrency.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020  -fsyntax-only -ast-dump -verify -pedantic %s | FileCheck %s
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+
+class Functor1 {
+public:
+  [[intel::max_concurrency(4)]] void operator()() const {}
+};
+
+[[intel::max_concurrency]] void foo() {} // expected-error {{'max_concurrency' attribute takes one argument}}
+
+class Functor2 {
+public:
+  void operator()() const {
+    foo();
+  }
+};
+
+template <int NT>
+class Functor3 {
+public:
+  [[intel::max_concurrency(NT)]] void operator()() const {}
+  // expected-error@+1 {{'max_concurrency' attribute only applies to 'for', 'while', 'do' statements, and functions}}
+  [[intel::max_concurrency(2)]] int a[10];
+};
+
+// expected-error@+1 {{'max_concurrency' attribute takes one argument}}
+[[intel::max_concurrency(3, 3)]] void goo() {}
+
+class Functor4 {
+public:
+  void operator() () const {
+    goo();
+  }
+};
+
+// expected-error@+1 {{'max_concurrency' attribute requires a positive integral compile time constant expression}}
+[[intel::max_concurrency(-1)]] void bar() {}
+class Functor5 {
+public:
+  void operator() () const {
+    bar();
+  }
+};
+
+// expected-error@+1 {{integral constant expression must have integral or unscoped enumeration type, not 'const char [16]'}}
+[[intel::max_concurrency("numberofthreads")]] void zoo() {}
+
+template <int NT>
+[[intel::max_concurrency(NT)]] void func() {}
+
+[[intel::max_concurrency(8)]] void dup();
+[[intel::max_concurrency(9)]] void dup() {} // expected-error {{duplicate Intel FPGA function attribute 'max_concurrency'}}
+
+int main() {
+  queue q;
+
+  q.submit([&](handler &h) {
+    Functor1 f1;
+    h.single_task<class kernel_name1>(f1);
+
+    Functor2 f2;
+    h.single_task<class kernel_name2>(f2);
+
+    h.single_task<class kernel_name3>(
+      []() [[intel::max_concurrency(3)]]{}); // expected-error{{'max_concurrency' attribute cannot be applied to types}}
+
```

Comment:
We need to add let SupportsNonconformingLambdaSyntax = 1; inside Attr.td file since this attribute will apply on lambda i believe. @AaronBallman, what do you think?

We refactored the lambda attribute logic here:
https://github.com/intel/llvm/pull/3381: Move the lambda attribute hack logic into tablegen

Reply:
Is `[[intel::max_concurrency]]` specified by the SYCL standard (I didn't think it was)? The comments I added in #3381 in Attr.td include: `New attributes should not set this bit unless the attribute is required by the SYCL specification.` -- basically, we don't want things to opt into that behavior unless there's a strong reason for it because that behavior isn't conforming to C++.

Reply:
Thanks for the explanation.

Reply:
There's an argument to be made that we should make all the SYCL-related attributes consistent in this matter, but I'm trying to push back a little bit because I don't think we should try to upstream nonconforming extensions, so any attribute we mark that way is one I think we're marking as "not for upstreaming". That said, I have no idea what the feature request authors' needs are either.

Reply:
As far as I can tell max_concurrency is an Intel's attribute, so I suppose it should be marked as "not for upstreaming".

Reply:
We may want to think about coming up with some syntactic distinction between attributes we intend to upstream and ones we don't, but I'd want that to be a different feature than "make this attribute nonconforming".

My recommendation for this specific case is to leave it as an error for the moment, but put a comment above the test case that says something along the lines of `Intentionally remains an error until user need requires us to allow the attribute in this position. See the comments about SupportsNonconformingLambdaSyntax in Attr.td for details.`